src/C4EnergyBars.cpp

```diff
@@ -0,0 +1,586 @@
+{
+	if (!graphics) {
+		LogF("DefineEnergyBars Parameter 1 map expected but got: %s", C4Value(graphics).GetDataString().getData());
+		return nullptr;
+	}
+
+	if (!definition) {
+		LogF("DefineEnergyBars Parameter 2 array expected but got: %s", C4Value(definition).GetDataString().getData());
+		return nullptr;
+	}
+
+	int32_t value_index = 0;
+	C4EnergyBarsDef::Gfxs gfx;
+	C4EnergyBarsDef::Bars bars;
+	C4EnergyBarsDef::Names names;
+	bool success = ProcessGraphics(*graphics, gfx) && ProcessGroup(value_index, gfx, *definition, bars, true) && C4EnergyBarsDef::PopulateNamesFromValues(bars, names);
+	if(success) {
+		auto def = UniqueifyDefinition(new C4EnergyBarsDef(gfx, bars, names));
+		return Instantiate(def);
+	}
+	return nullptr;
+}
+
+bool C4EnergyBarsUniquifier::ProcessGraphics(C4ValueHash &map, C4EnergyBarsDef::Gfxs &gfx)
+{
+	auto amount = C4VString("amount");
+	auto scale  = C4VString("scale");
+
+	C4ValueHash::Iterator end = map.end();
+	for (C4ValueHash::Iterator it = map.begin(); it != end; ++it) {
+		C4Value key = (*it).first;
+		auto _key = key.getStr()->Data.getData();
+
+		C4Value val = (*it).second;
+		auto *_val = val.getMap();
+		auto &m = *_val;
+
+		C4Value file = m[C4VString("file")];
+		auto _file = _key;
+		if (file != C4VNull) _file = file.getStr()->Data.getData();
+
+		int32_t _amount = m[amount].getInt();
+		int32_t _scale  = m[scale].getInt();
+		if (_amount == 0) _amount = 1;
+		if (_scale == 0) _scale = 100;
+
+		auto it_success = gfx.emplace(_key, C4EnergyBarsDef::Gfx(_key, _file, _amount, _scale));
+		if (!it_success.second) {
+			LogF("DefineEnergyBars %s duplicate key in gfx description ", _key);
+			return false;
+		}
+	}
+
+	return true;
+}
+
+bool C4EnergyBarsUniquifier::ProcessGroup(int32_t &value_index, C4EnergyBarsDef::Gfxs &graphics, const C4ValueArray &group, C4EnergyBarsDef::Bars &bars, bool advanceAlways)
+{
+	int32_t size = group.GetSize();
+	for (int32_t i = 0; i < size; ++i) {
+		C4Value element = group[i];
+		switch (element.GetType()) {
+		case C4V_Map:
+		{
+			auto *map = element.getMap();
+			if (map) {
+				if(!ProcessEnergyBar(value_index, graphics, *map, bars, advanceAlways || i == size-1)) return false;
+			} else {
+				LogF("DefineEnergyBars got unexpected value: %s", element.GetDataString().getData());
+				return false;
+			}
+			break;
+		}
+		case C4V_Array:
+			if(advanceAlways) {
+				auto *array = element.getArray();
+				if (array) {
+					if(!ProcessGroup(value_index, graphics, *array, bars, false)) return false;
+				} else {
+					LogF("DefineEnergyBars got unexpected value: %s", element.GetDataString().getData());
+					return false;
+				}
+			} else {
+				LogF("DefineEnergyBars groups in groups are not allowed: %s", element.GetDataString().getData());
+				return false;
+			}
+			break;
+
+		default:
+			LogF("DefineEnergyBars array or map expected but got: %s", element.GetDataString().getData());
+			return false;
+			break;
```

Comment:
unnecessary

Reply:
Do you mean it should just hit the runtime error and that error will be shown to the user?
(I am assuming that there already is a catch somewhere that converts this to a script exception, but have to check, whether that's the case)

Reply:
Sorry, it seems like somehow it often marked more than I wanted to.
I only meant the break statement.

The error handling in itself is part of the bigger question.

Reply:
log / return / break at that position replaced with exception